crates/ide_assists/src/utils/gen_trait_fn_body.rs

```diff
@@ -149,16 +149,93 @@ fn gen_debug_impl(adt: &ast::Adt, func: &ast::Fn) -> Option<()> {
             let mut arms = vec![];
             for variant in list.variants() {
                 let name = variant.name()?;
-                let variant_name =
-                    make::path_pat(make::ext::path_from_idents(["Self", &format!("{}", name)])?);
-
+                let variant_name = make::ext::path_from_idents(["Self", &format!("{}", name)])?;
                 let target = make::expr_path(make::ext::ident_path("f").into());
-                let fmt_string = make::expr_literal(&(format!("\"{}\"", name))).into();
-                let args = make::arg_list(vec![target, fmt_string]);
-                let macro_name = make::expr_path(make::ext::ident_path("write"));
-                let macro_call = make::expr_macro_call(macro_name, args);
 
-                arms.push(make::match_arm(Some(variant_name.into()), None, macro_call.into()));
+                match variant.field_list() {
+                    Some(ast::FieldList::RecordFieldList(list)) => {
+                        let mut pats = vec![];
```

Comment:
Nit (because this code isn't really perf-sensitive): I think this allocation could be avoided by using iterators.

Reply:
You're not wrong, but we'd need to:

1. Iterate over all fields twice; I believe this in itself already allocates
2. We wouldn't be able to use `.into_iter().map` since we're using `?` to perform early returns in the loop. We could use a block to create an `impl Iterator` (I think?), but at that point we're sacrificing so much readability for perf that I don't think it's worth it.

Instead I propose we convert this to a constant number of allocations by pre-allocating enough space once. This can be done by doing something like:

```rust
let iter = list.fields();
let capacity = iter.size_hint().1.unwrap_or(0);
let mut pats = Vec::with_capacity(capacity);
for field in iter { ... }
```

Reply:
It's fine to alloc here for simplicity since this is in an assist invocation, not need to heavily optimize for performance.

Reply:
Actually even the pre-allocation code seems too much compared to just calling `vec![]`. Because this code isn't particularly sensitive to performance I think the best solution is to leave it as-is for now.